recon/src/protocol.rs

```diff
@@ -619,38 +660,25 @@ where
         Ok(())
     }
     // The remote is missing all keys in the range send them over.
-    async fn process_remote_missing_range(
+    fn process_remote_missing_ranges(
         &mut self,
-        range: &RangeHash<R::Key, R::Hash>,
-    ) -> Result<()> {
-        // TODO: This logic has two potential failure modes we need to test them
-        // 1. We allocate memory of all keys in the range, this can be very large.
-        // 2. We spend a lot of time writing out to the stream but not reading from the stream.
-        //    This can be a potential deadlock if both side enter this method for a large amount of
-        //    keys at the same time.
-
-        let keys = self
-            .recon
-            .range(range.first.clone(), range.last.clone(), 0, usize::MAX)
-            .await?;
-        for key in keys {
-            if let Some(value) = self.recon.value_for_key(key.clone()).await? {
-                self.stream
-                    .send((self.value_resp_fn)(
-                        self.sync_id.clone(),
-                        Value { key, value },
-                    ))
+        ranges: Vec<RangeHash<R::Key, R::Hash>>,
+    ) -> impl Stream<Item = Result<Value<R::Key>>> {
+        let recon = self.recon.clone();
+        try_stream! {
+            for range in ranges {
+                // TODO this holds all keys in memory
+                // Paginate or otherwise stream the keys and values back out
+                let keys = recon
+                    .range(range.first, range.last, 0, usize::MAX)
                     .await?;
+                for key in keys {
+                    if let Some(value) = recon.value_for_key(key.clone()).await? {
```

Comment:
If there isn't a value should we be concerned at all now that we should discover them together or am I forgetting about something?

Reply:
Good call, going to convert this into an error.